### PR TITLE
Engine support for remote transforms

### PR DESCRIPTION
**Pull request by Frassle**
# Description

This adds support to the engine for "remote transformations".

A transform is "remote" because it is being invoked via the engine on receiving a resource registration, rather than being ran locally in process before sending a resource registration. These transforms can also span multiple process boundaries, e.g. a transform function in a user program, then a transform function in a component library, both running for a resource registered by another component library.

The underlying new feature here is the idea of a `Callback`. The expectation is we're going to use callbacks for multiple features so these are _not_ defined in terms of transformations. A callback is an untyped byte array (usually will be a protobuf message), plus an address to define which server should be invoked to do the callback, and a token to identify it.

A language sdk can start up and serve a `Callbacks` service, keep a mapping of tokens to in-process functions (currently just using UUID's for this), and then pass that service address and token to the engine to be invoked later on.

The engine uses these callbacks to track transformations callbacks per resource, and on a new resource registrations invokes each relevant callback with the resource properties and options, having new properties and options returned that are then passed to the next relevant transform callback until all have been called and the engine has the final resource state and options to use.

## Checklist

- [x] I have run `make tidy` to update any new dependencies
- [x] I have run `make lint` to verify my code passes the lint check
  - [x] I have formatted my code using `gofumpt`

- [x] I have added tests that prove my fix is effective or that my feature works

- [x] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Cloud API version

### CODE REVIEW

File: pkg/resource/deploy/source_eval.go [hunk 1]
```diff
@@ -1248,10 +1387,129 @@ func (rm *resmon) RegisterResource(ctx context.Context,
...
+		props = newProps
+		opts = newOpts
+	}
+	// Lookup our parents transformations and run those
+	err = func() error {
+		// Function exists to scope the lock
+		rm.resourceTransformationsLock.Lock()
+		defer rm.resourceTransformationsLock.Unlock()
+		rm.resGoalsLock.Lock()
+		defer rm.resGoalsLock.Unlock()
```
Comment (tgummerer) [inline pkg/resource/deploy/source_eval.go:1473]: Would it make sense to have these as RW mutexes instead?  If I understand correctly this lock is only for reading from `rm.resourceTransformations`, not to make sure only one transform call is happening at a time?

File: sdk/go/common/slice/slice.go [hunk 2]
```diff
@@ -23,3 +23,27 @@ func Prealloc[T any](capacity int) []T {
 	}
 	return make([]T, 0, capacity)
 }
+
+// Map applies the given function to each element of the given slice and returns a new slice with the results.
+func Map[T, U any](s []T, f func(T) U) []U {
+	r := make([]U, len(s))
```
Comment (tgummerer) [inline sdk/go/common/slice/slice.go:29]:
```suggestion
	r := Prealloc[U](len(s))
```

Is there  a reason not to use the Prealloc function here?

**Review by tgummerer — COMMENTED** [review]
I went through this, and it all looks reasonable to me.  I'm not super familiar with this code yet though, so I'd appreciate if someone else could take a look as well.

File: pkg/resource/deploy/source_eval.go (hunk 1 above)
Comment (Frassle) [reply pkg/resource/deploy/source_eval.go:1473]: Yeh I think we could

File: changelog/pending/20240129--engine--add-support-for-remote-transformations-to-the-engine.yaml [hunk 3]
```diff
@@ -0,0 +1,4 @@
+changes:
+- type: feat
+  scope: engine
+  description: Add support for remote transformations to the engine.
```
Comment (justinvp) [inline changelog/pending/20240129--engine--add-support-for-remote-transformations-to-the-engine.yaml:4]:
```suggestion
  description: Add support for remote transforms to the engine.
```

File: pkg/engine/lifecycletest/transformation_test.go
Comment (justinvp) [inline pkg/engine/lifecycletest/transformation_test.go:1]: Can we add a test or two that uses transforms from a provider's `ConstructF`? Maybe add transform from program and show that it transforms resources created in `Construct`. And probably a test going the other way. Adding a transform from the component and then having it affect subsequent resources created in the program.

File: pkg/resource/deploy/source_eval.go [hunk 4]
```diff
@@ -732,6 +791,8 @@ func (rm *resmon) SupportsFeature(ctx context.Context,
 		hasSupport = true
 	case "deletedWith":
 		hasSupport = true
+	case "transformations":
```
Comment (justinvp) [inline pkg/resource/deploy/source_eval.go:794]: As we discussed:

```suggestion
	case "transforms":
```

File: proto/pulumi/resource.proto [hunk 5]
```diff
@@ -169,3 +174,36 @@ message ResourceInvokeRequest {
...
+    repeated string additional_secret_outputs = 13;
+    map<string, string> providers = 14;
+    map<string, bytes> plugin_checksums = 15;
+}
+
+message TransformationRequest {
+    string type = 1; // the type of the resource.
+    string name = 2; // the name of the resource.
+    bool custom = 3; // true if the resource is a custom resource, else it's a component resource.
+    string parent = 4; // the parent of the resource, this can't be changed by the transformation.
```
[the remark this answers is not in the document]
Comment (justinvp) [reply proto/pulumi/resource.proto:201]: Ohh, I think I know: we don't want to allow changing the parent from the transform.

File: pkg/engine/lifecycletest/transformation_test.go
Comment (justinvp) [reply pkg/engine/lifecycletest/transformation_test.go:1]: We could also use a test that passes all the various resource options to a resource and ensures we get the expected values in the transform.

**Review by justinvp — COMMENTED** [review]
Looks good overall

File: proto/pulumi/resource.proto (hunk 5 above)
[the remark this answers is not in the document]
Comment (Frassle) [reply proto/pulumi/resource.proto:201]: Yup

File: sdk/go/pulumi/mocks.go [hunk 6]
```diff
@@ -262,6 +262,12 @@ func (m *mockMonitor) RegisterResourceOutputs(ctx context.Context, in *pulumirpc
 	return &emptypb.Empty{}, nil
 }
 
+func (m *mockMonitor) RegisterStackTransformation(ctx context.Context, in *pulumirpc.Callback,
+	opts ...grpc.CallOption,
+) (*emptypb.Empty, error) {
+	panic("not implemented")
```
[the remark this answers is not in the document]
Comment (Frassle) [reply sdk/go/pulumi/mocks.go:268]: Yup we should

File: pkg/engine/lifecycletest/transformation_test.go
Comment (Frassle) [reply pkg/engine/lifecycletest/transformation_test.go:1]: Yup we should, I noticed a few options I didn't copy across in the first case.

Comment (Frassle) [reply pkg/engine/lifecycletest/transformation_test.go:1]: Callback test added, just need to write the options test up.

**Review by julienp — APPROVED** [review]

File: pkg/resource/deploy/source_eval.go (hunk 4 above)
Comment (justinvp) [inline pkg/resource/deploy/source_eval.go:794]: I thought we were going to change this to "transforms"?

File: proto/pulumi/resource.proto [hunk 7]
```diff
@@ -134,6 +137,8 @@ message RegisterResourceRequest {
     bool aliasSpecs = 28;
 
     SourcePosition sourcePosition = 29;    // the optional source position of the user code that initiated the register.
+
+    repeated Callback transforms = 39; // a list of transforms to apply to the resource before registering it.
```
Comment (justinvp) [inline proto/pulumi/resource.proto:141]: Should this be 31? Does it matter?

**Review by justinvp — APPROVED** [review]
LGTM otherwise

File: pkg/resource/deploy/source_eval.go (hunk 4 above)
Comment (Frassle) [reply pkg/resource/deploy/source_eval.go:794]: Yup, done.

**Pull request merged**